plugins/cache_promote/lru_policy.cc

```diff
@@ -72,101 +112,130 @@ LRUPolicy::doPromote(TSHttpTxn txnp)
+    incrementStat(_lru_miss_id, 1);
     if (_list_size >= _buckets) {
       TSDebug(PLUGIN_NAME, "repurposing last LRUHash entry");
       _list.splice(_list.begin(), _list, --_list.end());
-      _map.erase(&(_list.begin()->first));
-      incrementStat(lru_vacated_id, 1);
+      _map.erase(&(std::get<0>(*_list.begin()))); // Get the hash from the first list element
+      incrementStat(_lru_vacated_id, 1);
     } else if (_freelist_size > 0) {
       TSDebug(PLUGIN_NAME, "reusing LRUEntry from freelist");
       _list.splice(_list.begin(), _freelist, _freelist.begin());
       --_freelist_size;
       ++_list_size;
-      incrementStat(lru_size_id, 1);
-      decrementStat(freelist_size_id, 1);
+      incrementStat(_lru_size_id, 1);
+      decrementStat(_freelist_size_id, 1);
     } else {
       TSDebug(PLUGIN_NAME, "creating new LRUEntry");
       _list.push_front(NULL_LRU_ENTRY);
       ++_list_size;
-      incrementStat(lru_size_id, 1);
+      incrementStat(_lru_size_id, 1);
     }
     // Update the "new" LRUEntry and add it to the hash
-    _list.begin()->first          = hash;
-    _list.begin()->second         = 1;
-    _map[&(_list.begin()->first)] = _list.begin();
+    *_list.begin()                       = {hash, 1, 0};
+    _map[&(std::get<0>(*_list.begin()))] = _list.begin();
   }
 
   TSMutexUnlock(_lock);
 
+  // If we didn't promote, and we want to count bytes, save away the calculated hash for later use
+  if (false == ret && countBytes()) {
+    TSUserArgSet(txnp, TXN_ARG_IDX, static_cast<void *>(new LRUHash(hash)));
+  } else {
+    TSUserArgSet(txnp, TXN_ARG_IDX, nullptr);
+  }
+
   return ret;
 }
 
+void
+LRUPolicy::addBytes(TSHttpTxn txnp)
+{
+  LRUHash *hash = static_cast<LRUHash *>(TSUserArgGet(txnp, TXN_ARG_IDX));
+
+  if (hash) {
+    LRUMap::iterator map_it;
+
+    // We have to hold the lock across all list and hash access / updates
+    TSMutexLock(_lock);
+    map_it = _map.find(hash);
+    if (_map.end() != map_it) {
+      TSMBuffer resp;
+      TSMLoc resp_hdr;
+
+      if (TS_SUCCESS == TSHttpTxnServerRespGet(txnp, &resp, &resp_hdr)) {
+        TSMLoc field_loc = TSMimeHdrFieldFind(resp, resp_hdr, TS_MIME_FIELD_CONTENT_LENGTH, TS_MIME_LEN_CONTENT_LENGTH);
+
+        if (field_loc) {
+          auto &[val_key, val_hits, val_bytes] = *(map_it->second);
+          int64_t cl                           = TSMimeHdrFieldValueInt64Get(resp, resp_hdr, field_loc, -1);
+
+          // This is beacuse compilers before gcc 8 aren't smart enough to ignore the unused structured bindings
+          (void)val_key, (void)val_hits;
+
+          val_bytes += cl;
+          TSDebug(PLUGIN_NAME, "Added %" PRId64 " bytes for LRU entry", cl);
+          TSHandleMLocRelease(resp, resp_hdr, field_loc);
+        }
+        TSHandleMLocRelease(resp, TS_NULL_MLOC, resp_hdr);
+      }
+    }
+    TSMutexUnlock(_lock);
+
+    // Delete the hash, and remove the pointer from the TXN user arg slot (to be safe)
+    delete hash;
```

Comment:
What if the transaction fails before it gets to the hook where this is called?  I think you need to trigger on the TXN_CLOSE hook and delete this there.

Reply:
I know it's hard for men of our age not to leak but we have to try.

Reply:
Consider moving Cleanup.h to tscore.  It has the logic you seem to need.

https://github.com/apache/trafficserver/blob/8715068f83f35bbe44bf6bec236f7e67995e865f/plugins/xdebug/xdebug.cc#L63

Reply:
> What if the transaction fails before it gets to the hook where this is called? I think you need to trigger on the TXN_CLOSE hook and delete this there.

Ok, will take a look, thanks!

Reply:
And death to ATS CPP API

Reply:
Don't throw out the baby with the bathwater:  https://github.com/apache/trafficserver/blob/3d584af796bad1e9e3c03b2af5485b630ffedafb/plugins/xdebug/Cleanup.h#L21